File: examples/websocket_client.py
#!/usr/bin/env python3
"""
WebSocket Client Example for RustyDB

This example demonstrates how to connect to RustyDB using WebSocket
for real-time query execution and streaming results.

Requirements:
    pip install websockets asyncio

Usage:
    python websocket_client.py
"""

import asyncio
import json
import uuid
from datetime import datetime, timezone
from typing import Optional, Dict, Any
import websockets
from websockets.client import WebSocketClientProtocol


class RustyDBWebSocketClient:
    """WebSocket client for RustyDB with automatic reconnection and error handling."""

    def __init__(
        self,
        url: str = "ws://localhost:8080/api/v1/stream",
        auth_token: Optional[str] = None,
        auto_reconnect: bool = True,
        max_reconnect_attempts: int = 5,
    ):
        """
        Initialize the WebSocket client.

        Args:
            url: WebSocket server URL
            auth_token: JWT authentication token (optional)
            auto_reconnect: Enable automatic reconnection
            max_reconnect_attempts: Maximum number of reconnection attempts
        """
        self.url = url
        self.auth_token = auth_token
        self.auto_reconnect = auto_reconnect
        self.max_reconnect_attempts = max_reconnect_attempts
        self.websocket: Optional[WebSocketClientProtocol] = None
        self.reconnect_count = 0
        self.authenticated = False

    async def connect(self) -> None:
        """Establish WebSocket connection to RustyDB."""
        print(f"Connecting to: {self.url}")

        try:
            self.websocket = await websockets.connect(self.url)
            print("✓ Connected successfully")

            # Authenticate if token is provided
            if self.auth_token:
                await self.authenticate()

            self.reconnect_count = 0

        except Exception as e:
            print(f"✗ Connection failed: {e}")
            raise

    async def authenticate(self) -> bool:
        """Authenticate with the server using JWT token."""
        if not self.auth_token:
            print("⚠ No authentication token provided")
            return False

        auth_message = {
            "type": "auth",
            "payload": {"token": self.auth_token},
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "id": str(uuid.uuid4()),
        }

        await self.send_message(auth_message)
        print("✓ Sent authentication message")

        # Wait for authentication response
        response = await self.receive_message()
        if response and response["type"] == "auth_success":
            print("✓ Authentication successful")
            self.authenticated = True
            return True
        elif response and response["type"] == "auth_error":
            print(f"✗ Authentication failed: {response.get('payload', {})}")
            self.authenticated = False
            return False

        return False

    async def send_message(self, message: Dict[str, Any]) -> None:
        """Send a message to the server."""
        if not self.websocket:
            raise RuntimeError("Not connected to server")

        message_json = json.dumps(message)
        await self.websocket.send(message_json)

    async def receive_message(self, timeout: Optional[float] = 5.0) -> Optional[Dict[str, Any]]:
        """Receive a message from the server with optional timeout."""
        if not self.websocket:
            raise RuntimeError("Not connected to server")

        try:
            if timeout:
                message_text = await asyncio.wait_for(
                    self.websocket.recv(), timeout=timeout
                )
            else:
                message_text = await self.websocket.recv()

            return json.loads(message_text)

        except asyncio.TimeoutError:
            print("⚠ Timeout waiting for response")
            return None
        except json.JSONDecodeError as e:
            print(f"✗ Failed to parse JSON: {e}")
            return None

    async def execute_query(
        self, sql: str, params: Optional[list] = None, streaming: bool = True
    ) -> Optional[Dict[str, Any]]:
        """
        Execute a SQL query and receive results.

        Args:
            sql: SQL query to execute
            params: Query parameters (optional)
            streaming: Enable streaming results

        Returns:
            Query response dictionary
        """
        query_message = {
            "type": "query",
            "payload": {
                "sql": sql,
                "params": params or [],
                "streaming": streaming,
            },
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "id": str(uuid.uuid4()),
        }

        await self.send_message(query_message)
        return await self.receive_message()

    async def ping(self) -> bool:
        """Send a ping message and wait for pong response."""
        ping_message = {
            "type": "ping",
            "payload": {},
            "timestamp": datetime.now(timezone.utc).isoformat(),
        }

        await self.send_message(ping_message)

        response = await self.receive_message(timeout=5.0)
        return response is not None and response["type"] == "pong"

    async def close(self) -> None:
        """Close the WebSocket connection."""
        if self.websocket:
            await self.websocket.close()
            print("✓ Connection closed")
            self.websocket = None
            self.authenticated = False

    async def __aenter__(self):
        """Async context manager entry."""
        await self.connect()
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Async context manager exit."""
        await self.close()


async def example_1_basic_connection():
    """Example 1: Basic connection and authentication."""
    print("\n" + "=" * 50)
    print("Example 1: Basic Connection")
    print("=" * 50 + "\n")

    # Create client without authentication
    async with RustyDBWebSocketClient() as client:
        print("✓ Connection established and closed successfully")


async def example_2_authenticated_connection():
    """Example 2: Authenticated connection."""
    print("\n" + "=" * 50)
    print("Example 2: Authenticated Connection")
    print("=" * 50 + "\n")

    # In production, you would get this token from the login endpoint
    auth_token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9..."

    async with RustyDBWebSocketClient(auth_token=auth_token) as client:
        if client.authenticated:
            print("✓ Authenticated connection ready for queries")
        else:
            print("⚠ Connected but not authenticated")


async def example_3_query_execution():
    """Example 3: Execute queries and receive results."""
    print("\n" + "=" * 50)
    print("Example 3: Query Execution")
    print("=" * 50 + "\n")

    async with RustyDBWebSocketClient() as client:
        # Execute a simple query
        sql = "SELECT name, age FROM users WHERE age > 18 ORDER BY name LIMIT 5"
        print(f"Executing query: {sql}")

        response = await client.execute_query(sql)

        if response and response["type"] == "data":
            payload = response["payload"]
            print(f"✓ Query executed successfully")
            print(f"  Status: {payload.get('status')}")
            print(f"  Columns: {payload.get('columns', [])}")
            print(f"  Rows returned: {len(payload.get('rows', []))}")

            # Print results
            rows = payload.get("rows", [])
            if rows:
                print("\n  Results:")
                for i, row in enumerate(rows, 1):
                    print(f"    Row {i}: {row}")

        elif response and response["type"] == "error":
            print(f"✗ Query error: {response['payload'].get('message')}")


async def example_4_heartbeat():
    """Example 4: Heartbeat (ping/pong)."""
    print("\n" + "=" * 50)
    print("Example 4: Heartbeat")
    print("=" * 50 + "\n")

    async with RustyDBWebSocketClient() as client:
        print("Sending ping...")
        if await client.ping():
            print("✓ Received pong response")
        else:
            print("✗ No pong response received")


async def example_5_error_handling():
    """Example 5: Error handling for invalid queries."""
    print("\n" + "=" * 50)
    print("Example 5: Error Handling")
    print("=" * 50 + "\n")

    async with RustyDBWebSocketClient() as client:
        # Execute an invalid query
        sql = "SELECT * FROM nonexistent_table"
        print(f"Executing invalid query: {sql}")

        response = await client.execute_query(sql)

        if response and response["type"] == "error":
            print("✓ Received expected error response")
            print(f"  Error message: {response['payload'].get('message')}")
            print(f"  Error code: {response['payload'].get('code')}")
        else:
            print("⚠ Expected error but got different response")


async def example_6_multiple_queries():
    """Example 6: Execute multiple queries sequentially."""
    print("\n" + "=" * 50)
    print("Example 6: Multiple Queries")
    print("=" * 50 + "\n")

    async with RustyDBWebSocketClient() as client:
        queries = [
            "SELECT COUNT(*) FROM users",
            "SELECT AVG(age) FROM users",
            "SELECT MAX(created_at) FROM users",
        ]

        for i, sql in enumerate(queries, 1):
            print(f"\nQuery {i}: {sql}")
            response = await client.execute_query(sql)

            if response and response["type"] == "data":
                payload = response["payload"]
                rows = payload.get("rows", [])
                if rows:
                    print(f"  Result: {rows[0]}")
            else:
                print(f"  Error: {response}")


async def example_7_reconnection():
    """Example 7: Automatic reconnection on failure."""
    print("\n" + "=" * 50)
    print("Example 7: Reconnection Logic")
    print("=" * 50 + "\n")

    client = RustyDBWebSocketClient(
        auto_reconnect=True,
        max_reconnect_attempts=3
    )

    max_retries = 3
    retry_count = 0
    retry_delay = 1.0

    while retry_count < max_retries:
        try:
            await client.connect()
            print("✓ Connected successfully")

            # Simulate some work
            await asyncio.sleep(1)

            break

        except Exception as e:
            retry_count += 1
            if retry_count >= max_retries:
                print(f"✗ Max retries ({max_retries}) reached, giving up")
                break

            print(f"✗ Connection failed (attempt {retry_count}/{max_retries}): {e}")
            print(f"  Retrying in {retry_delay}s...")
            await asyncio.sleep(retry_delay)

            # Exponential backoff
            retry_delay = min(retry_delay * 2, 60.0)

    await client.close()


async def example_8_streaming_results():
    """Example 8: Handle streaming results from large queries."""
    print("\n" + "=" * 50)
    print("Example 8: Streaming Results")
    print("=" * 50 + "\n")

    async with RustyDBWebSocketClient() as client:
        # Execute a query that returns a large result set
        sql = "SELECT * FROM large_table LIMIT 1000"
        print(f"Executing query: {sql}")

        await client.send_message({
            "type": "query",
            "payload": {
                "sql": sql,
                "streaming": True,
            },
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "id": str(uuid.uuid4()),
        })

        # Receive streaming results
        total_rows = 0
        while True:
            response = await client.receive_message(timeout=10.0)

            if not response:
                print("⚠ No more messages")
                break

            if response["type"] == "data":
                payload = response["payload"]
                rows = payload.get("rows", [])
                total_rows += len(rows)
                print(f"  Received batch: {len(rows)} rows (total: {total_rows})")

                # Check if there are more results
                if not payload.get("has_more", False):
                    print("✓ All results received")
                    break

            elif response["type"] == "error":
                print(f"✗ Error: {response['payload'].get('message')}")
                break


async def main():
    """Run all examples."""
    print("=" * 50)
    print("RustyDB WebSocket Client Examples")
    print("=" * 50)

    # Run examples
    try:
        await example_1_basic_connection()
        await example_2_authenticated_connection()
        await example_3_query_execution()
        await example_4_heartbeat()
        await example_5_error_handling()
        await example_6_multiple_queries()
        await example_7_reconnection()
        await example_8_streaming_results()

        print("\n" + "=" * 50)
        print("All examples completed successfully!")
        print("=" * 50)

    except KeyboardInterrupt:
        print("\n\n✗ Interrupted by user")
    except Exception as e:
        print(f"\n\n✗ Error running examples: {e}")
        import traceback
        traceback.print_exc()


if __name__ == "__main__":
    # Run the async main function
    asyncio.run(main())

File: examples/websocket_client.rs
// # WebSocket Client Example
//
// This example demonstrates how to connect to RustyDB using WebSocket
// for real-time query execution and streaming results.

use tokio_tungstenite::{connect_async, tungstenite::protocol::Message};
use futures_util::{StreamExt, SinkExt};
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::time::Duration;

// WebSocket message structure
#[derive(Debug, Serialize, Deserialize)]
struct WebSocketMessage {
    #[serde(rename = "type")]
    message_type: String,
    payload: serde_json::Value,
    timestamp: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    id: Option<String>,
}

// Query response structure
#[derive(Debug, Deserialize)]
struct QueryResponse {
    status: String,
    #[serde(default)]
    rows: Vec<Vec<serde_json::Value>>,
    #[serde(default)]
    columns: Vec<String>,
    #[serde(default)]
    rows_affected: usize,
    #[serde(skip_serializing_if = "Option::is_none")]
    message: Option<String>,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("=== RustyDB WebSocket Client Example ===\n");

    // Example 1: Basic connection and authentication
    println!("Example 1: Basic Connection");
    println!("----------------------------");
    basic_connection_example().await?;

    // Example 2: Execute queries with streaming results
    println!("\nExample 2: Query Execution");
    println!("----------------------------");
    query_execution_example().await?;

    // Example 3: Heartbeat (ping/pong)
    println!("\nExample 3: Heartbeat");
    println!("----------------------------");
    heartbeat_example().await?;

    // Example 4: Error handling
    println!("\nExample 4: Error Handling");
    println!("----------------------------");
    error_handling_example().await?;

    println!("\n=== All examples completed ===");

    Ok(())
}

// Example 1: Basic connection and authentication
async fn basic_connection_example() -> Result<(), Box<dyn std::error::Error>> {
    // Connect to RustyDB WebSocket endpoint
    let url = "ws://localhost:8080/api/v1/stream";
    println!("Connecting to: {}", url);

    let (ws_stream, _) = connect_async(url).await?;
    println!("✓ Connected successfully");

    let (mut write, mut read) = ws_stream.split();

    // Send authentication message
    let auth_message = WebSocketMessage {
        message_type: "auth".to_string(),
        payload: json!({
            "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9..."
        }),
        timestamp: chrono::Utc::now().to_rfc3339(),
        id: Some(uuid::Uuid::new_v4().to_string()),
    };

    let auth_json = serde_json::to_string(&auth_message)?;
    write.send(Message::Text(auth_json)).await?;
    println!("✓ Sent authentication message");

    // Wait for authentication response
    if let Some(msg) = read.next().await {
        match msg? {
            Message::Text(text) => {
                let response: WebSocketMessage = serde_json::from_str(&text)?;
                match response.message_type.as_str() {
                    "auth_success" => println!("✓ Authentication successful"),
                    "auth_error" => {
                        println!("✗ Authentication failed: {:?}", response.payload);
                        return Err("Authentication failed".into());
                    }
                    _ => println!("  Received: {}", response.message_type),
                }
            }
            _ => println!("  Received non-text message"),
        }
    }

    // Close connection
    write.close().await?;
    println!("✓ Connection closed\n");

    Ok(())
}

// Example 2: Execute queries with streaming results
async fn query_execution_example() -> Result<(), Box<dyn std::error::Error>> {
    let url = "ws://localhost:8080/api/v1/stream";
    let (ws_stream, _) = connect_async(url).await?;
    println!("✓ Connected to RustyDB");

    let (mut write, mut read) = ws_stream.split();

    // Execute a query
    let query_message = WebSocketMessage {
        message_type: "query".to_string(),
        payload: json!({
            "sql": "SELECT name, age FROM users WHERE age > 18 ORDER BY name LIMIT 5",
            "params": [],
            "streaming": true
        }),
        timestamp: chrono::Utc::now().to_rfc3339(),
        id: Some("query-001".to_string()),
    };

    let query_json = serde_json::to_string(&query_message)?;
    write.send(Message::Text(query_json)).await?;
    println!("✓ Sent query: SELECT name, age FROM users...");

    // Receive query results
    if let Some(msg) = read.next().await {
        match msg? {
            Message::Text(text) => {
                let response: WebSocketMessage = serde_json::from_str(&text)?;

                if response.message_type == "data" {
                    let query_response: QueryResponse = serde_json::from_value(response.payload)?;

                    println!("✓ Query executed successfully");
                    println!("  Status: {}", query_response.status);
                    println!("  Columns: {:?}", query_response.columns);
                    println!("  Rows returned: {}", query_response.rows.len());

                    // Print results
                    if !query_response.rows.is_empty() {
                        println!("\n  Results:");
                        for (i, row) in query_response.rows.iter().enumerate() {
                            println!("    Row {}: {:?}", i + 1, row);
                        }
                    }
                } else if response.message_type == "error" {
                    let query_response: QueryResponse = serde_json::from_value(response.payload)?;
                    println!("✗ Query error: {}", query_response.message.unwrap_or_default());
                }
            }
            _ => println!("  Received non-text message"),
        }
    }

    // Close connection
    write.close().await?;
    println!("\n✓ Connection closed\n");

    Ok(())
}

// Example 3: Heartbeat (ping/pong)
async fn heartbeat_example() -> Result<(), Box<dyn std::error::Error>> {
    let url = "ws://localhost:8080/api/v1/stream";
    let (ws_stream, _) = connect_async(url).await?;
    println!("✓ Connected to RustyDB");

    let (mut write, mut read) = ws_stream.split();

    // Send ping message
    let ping_message = WebSocketMessage {
        message_type: "ping".to_string(),
        payload: json!({}),
        timestamp: chrono::Utc::now().to_rfc3339(),
        id: None,
    };

    let ping_json = serde_json::to_string(&ping_message)?;
    write.send(Message::Text(ping_json)).await?;
    println!("✓ Sent ping");

    // Wait for pong response with timeout
    match tokio::time::timeout(Duration::from_secs(5), read.next()).await {
        Ok(Some(msg)) => {
            match msg? {
                Message::Text(text) => {
                    let response: WebSocketMessage = serde_json::from_str(&text)?;
                    if response.message_type == "pong" {
                        println!("✓ Received pong");
                        println!("  Server timestamp: {}", response.timestamp);
                    } else {
                        println!("  Received: {}", response.message_type);
                    }
                }
                Message::Pong(_) => println!("✓ Received WebSocket pong"),
                _ => println!("  Received other message type"),
            }
        }
        Ok(None) => println!("✗ Connection closed"),
        Err(_) => println!("✗ Timeout waiting for pong"),
    }

    // Close connection
    write.close().await?;
    println!("\n✓ Connection closed\n");

    Ok(())
}

// Example 4: Error handling
async fn error_handling_example() -> Result<(), Box<dyn std::error::Error>> {
    let url = "ws://localhost:8080/api/v1/stream";
    let (ws_stream, _) = connect_async(url).await?;
    println!("✓ Connected to RustyDB");

    let (mut write, mut read) = ws_stream.split();

    // Execute an invalid query
    let invalid_query = WebSocketMessage {
        message_type: "query".to_string(),
        payload: json!({
            "sql": "SELECT * FROM nonexistent_table",
            "params": []
        }),
        timestamp: chrono::Utc::now().to_rfc3339(),
        id: Some("query-error-001".to_string()),
    };

    let query_json = serde_json::to_string(&invalid_query)?;
    write.send(Message::Text(query_json)).await?;
    println!("✓ Sent invalid query (expecting error)");

    // Receive error response
    if let Some(msg) = read.next().await {
        match msg? {
            Message::Text(text) => {
                let response: WebSocketMessage = serde_json::from_str(&text)?;

                if response.message_type == "error" {
                    println!("✓ Received expected error response");
                    println!("  Error details: {:?}", response.payload);

                    // Parse error details
                    if let Some(message) = response.payload.get("message") {
                        println!("  Error message: {}", message);
                    }
                    if let Some(code) = response.payload.get("code") {
                        println!("  Error code: {}", code);
                    }
                } else {
                    println!("  Unexpected response type: {}", response.message_type);
                }
            }
            _ => println!("  Received non-text message"),
        }
    }

    // Close connection
    write.close().await?;
    println!("\n✓ Connection closed\n");

    Ok(())
}

// Advanced Example: Reconnection logic with exponential backoff
#[allow(dead_code)]
async fn advanced_reconnection_example() -> Result<(), Box<dyn std::error::Error>> {
    let url = "ws://localhost:8080/api/v1/stream";
    let max_retries = 5;
    let mut retry_count = 0;
    let mut retry_delay = Duration::from_secs(1);

    loop {
        match connect_async(url).await {
            Ok((ws_stream, _)) => {
                println!("✓ Connected successfully");
                let (mut write, mut read) = ws_stream.split();

                // Handle messages
                while let Some(msg) = read.next().await {
                    match msg {
                        Ok(Message::Text(text)) => {
                            println!("Received: {}", text);
                        }
                        Ok(Message::Close(_)) => {
                            println!("Server closed connection");
                            break;
                        }
                        Err(e) => {
                            println!("Error receiving message: {}", e);
                            break;
                        }
                        _ => {}
                    }
                }

                // Connection closed, attempt reconnection
                retry_count = 0;
                retry_delay = Duration::from_secs(1);
            }
            Err(e) => {
                retry_count += 1;
                if retry_count > max_retries {
                    println!("✗ Max retries reached, giving up");
                    return Err(e.into());
                }

                println!(
                    "✗ Connection failed (attempt {}/{}): {}",
                    retry_count, max_retries, e
                );
                println!("  Retrying in {:?}...", retry_delay);

                tokio::time::sleep(retry_delay).await;

                // Exponential backoff
                retry_delay = std::cmp::min(
                    retry_delay * 2,
                    Duration::from_secs(60)
                );
            }
        }
    }
}

// Advanced Example: Multiple concurrent subscriptions
#[allow(dead_code)]
async fn multiple_subscriptions_example() -> Result<(), Box<dyn std::error::Error>> {
    let url = "ws://localhost:8080/api/v1/stream";
    let (ws_stream, _) = connect_async(url).await?;
    println!("✓ Connected to RustyDB");

    let (mut write, mut read) = ws_stream.split();

    // Send multiple queries
    let queries = vec![
        ("query-1", "SELECT * FROM users LIMIT 5"),
        ("query-2", "SELECT * FROM products LIMIT 5"),
        ("query-3", "SELECT * FROM orders LIMIT 5"),
    ];

    for (id, sql) in &queries {
        let query_message = WebSocketMessage {
            message_type: "query".to_string(),
            payload: json!({
                "sql": sql,
                "streaming": true
            }),
            timestamp: chrono::Utc::now().to_rfc3339(),
            id: Some(id.to_string()),
        };

        let query_json = serde_json::to_string(&query_message)?;
        write.send(Message::Text(query_json)).await?;
        println!("✓ Sent query: {}", id);
    }

    // Receive all responses
    let mut responses_received = 0;
    while responses_received < queries.len() {
        if let Some(msg) = read.next().await {
            match msg? {
                Message::Text(text) => {
                    let response: WebSocketMessage = serde_json::from_str(&text)?;
                    println!("✓ Received response for: {:?}", response.id);
                    responses_received += 1;
                }
                _ => {}
            }
        }
    }

    write.close().await?;
    println!("✓ All queries completed\n");

    Ok(())
}

// Helper: Format duration for display
#[allow(dead_code)]
fn format_duration(duration: Duration) -> String {
    let secs = duration.as_secs();
    let millis = duration.subsec_millis();
    format!("{}.{:03}s", secs, millis)
}
